preprocess.py

```diff
@@ -0,0 +1,205 @@
+    }
+
+
+def handle_ForkEvent(d):
+    return {
+        "username": d["actor"]["login"],
+        "repo": d["repo"]["name"],
+        "created_at": datetime.datetime.fromisoformat(d["created_at"]),
+    }
+
+
+
+conversions = {
+    "PushEvent": handle_PushEvent,
+    "CreateEvent": handle_CreateEvent,
+    "PullRequestEvent": handle_PullRequestEvent,
+    "IssueCommentEvent": handle_IssueCommentEvent,
+    "WatchEvent": handle_WatchEvent,
+    "ForkEvent": handle_ForkEvent,
+}
+
+
+def process_records(seq):
+    out = defaultdict(list)
+    for record in seq:
+        try:
+            record = json.loads(record)
+        except Exception:
+            continue
+
+        if record["type"] in conversions:
+            convert = conversions[record["type"]]
+            try:
+                result = convert(record)
+            except Exception as e:
+                logger.error(f"Failed to parse record: {e}")
+                continue
+
+            out[record["type"]].append(result)
+
+    return out
+
+
+def process_file(filename: str) -> dict[str, pd.DataFrame]:
+    with fsspec.open(
+        filename, compression="gzip" if filename.endswith("gz") else None
+    ) as f:
+        out = process_records(f)
+        out["Commits"] = list(toolz.concat(out["PushEvent"]))
+
+    del out["PushEvent"]
+    out = toolz.valmap(pd.DataFrame, out)
+    out = {
+        "comment": out["IssueCommentEvent"],
+        "pr": out["PullRequestEvent"],
+        "commit": out["Commits"],
+        "create": out["CreateEvent"],
+        "watch": out["WatchEvent"],
+        "fork": out["ForkEvent"],
+    }
+    return out
+
+
+def write_delta(tables: dict[str, pd.DataFrame]):
+    for table, df in tables.items():
+        outfile = OUTDIR / table
+        outfile.fs.makedirs(outfile.parent, exist_ok=True)
+        deltalake.write_deltalake(
+            outfile,
+            df,
+            mode="append",
+            storage_options=STORAGE_OPTIONS,
+        )
+
+def list_files(start, stop):
+    dates = pd.date_range(start, stop - datetime.timedelta(days=1), freq="d")
+    hours = [str(dt.date()) + f"-{h}.json.gz" for dt in dates for h in range(24)]
+    filenames = ["https://data.gharchive.org/" + hour for hour in hours]
+    return filenames
+
+
+def compact(table):
+    t = deltalake.DeltaTable(
+        OUTDIR / table,
+        storage_options=STORAGE_OPTIONS,
+    )
+    t.optimize.compact()
+    t.vacuum(retention_hours=0, enforce_retention_duration=False, dry_run=False)
+    print(f"Compacted {table} table")
+
+
+if __name__ == "__main__":
+    filenames = list_files(
+        start=datetime.date(2024, 1, 1),
+        stop=datetime.date(2024, 1, 10),
+    )
+
+    with Cluster() as cluster:
+        if not LOCAL:
+            cluster.send_private_envs({
+                "AWS_ACCESS_KEY_ID": os.environ["AWS_ACCESS_KEY_ID"],
+                "AWS_SECRET_ACCESS_KEY": os.environ["AWS_SECRET_ACCESS_KEY"],
+            })
```

Comment:
It's unfortunate this is needed, but delta needs these environment variables to be set